File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/sparse/CsrDiagMmBp.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  * See the NOTICE file distributed with this work for additional
 *  * information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.sparse;

import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Arrays;
import java.util.List;

/**
 * Exact backward pass for {@link CsrDiagMm}: {@code out[e] = dl[i_e] * aValues[e] * dr[j_e]}.
 *
 * <p>Given the upstream gradient {@code gradOut[nnz]} w.r.t. the forward output values
 * and the original forward inputs, computes:
 * <pre>
 *   dAValues[e]  = gradOut[e] * dl[i_e] * dr[j_e]
 *   ddl[i]       = sum_{e in row i}    gradOut[e] * aValues[e] * dr[j_e]
 *   ddr[j]       = sum_{e: col j_e==j} gradOut[e] * aValues[e] * dl[i_e]
 * </pre>
 *
 * <p>C++ op name: {@code csr_diag_mm_bp}
 *
 * <p><b>Inputs (6 arrays):</b>
 * <ol>
 *   <li>{@code aValues}  – 1D [nnz],    floating dtype — original non-zero values of A</li>
 *   <li>{@code aColIdx}  – 1D [nnz],    INT32 — column indices of A</li>
 *   <li>{@code aRowPtr}  – 1D [rows+1], INT32 — row pointers of A</li>
 *   <li>{@code dl}       – 1D [rows],   floating dtype — left diagonal</li>
 *   <li>{@code dr}       – 1D [cols],   floating dtype — right diagonal</li>
 *   <li>{@code gradOut}  – 1D [nnz],    floating dtype — upstream gradient w.r.t. outValues</li>
 * </ol>
 *
 * <p><b>Integer arguments (IArgs):</b> {@code rows}, {@code cols}.
 *
 * <p><b>Outputs (3 arrays):</b>
 * <ol>
 *   <li>{@code dAValues} – 1D [nnz],  same floating dtype as {@code aValues}</li>
 *   <li>{@code ddl}      – 1D [rows], same floating dtype as {@code dl}</li>
 *   <li>{@code ddr}      – 1D [cols], same floating dtype as {@code dr}</li>
 * </ol>
 *
 * <p>This op is forward-only (no {@code doDiff}); it is the gradient leaf created by
 * {@link CsrDiagMm#doDiff}.
 */
public class CsrDiagMmBp extends DynamicCustomOp {

    /** No-arg constructor required for op-registry reflection. */
    public CsrDiagMmBp() {}

    /**
     * Eager (INDArray) constructor.
     *
     * @param aValues  1D [nnz]    original non-zero values of A (floating dtype)
     * @param aColIdx  1D [nnz]    INT32 column indices of A
     * @param aRowPtr  1D [rows+1] INT32 row pointers of A
     * @param dl       1D [rows]   left diagonal vector (floating dtype)
     * @param dr       1D [cols]   right diagonal vector (floating dtype)
     * @param gradOut  1D [nnz]    upstream gradient w.r.t. outValues (floating dtype)
     * @param rows     number of rows in the sparse matrix
     * @param cols     number of columns in the sparse matrix
     */
    public CsrDiagMmBp(INDArray aValues, INDArray aColIdx, INDArray aRowPtr,
                        INDArray dl, INDArray dr, INDArray gradOut,
                        long rows, long cols) {
        super(new INDArray[]{aValues, aColIdx, aRowPtr, dl, dr, gradOut}, null);
        addIArgument(rows, cols);
    }

    /**
     * SameDiff (symbolic) constructor.
     *
     * @param sd       the SameDiff graph
     * @param aValues  SDVariable [nnz]    original non-zero values of A
     * @param aColIdx  SDVariable [nnz]    INT32 column indices of A
     * @param aRowPtr  SDVariable [rows+1] INT32 row pointers of A
     * @param dl       SDVariable [rows]   left diagonal vector
     * @param dr       SDVariable [cols]   right diagonal vector
     * @param gradOut  SDVariable [nnz]    upstream gradient w.r.t. outValues
     * @param rows     number of rows
     * @param cols     number of columns
     */
    public CsrDiagMmBp(SameDiff sd,
                        SDVariable aValues, SDVariable aColIdx, SDVariable aRowPtr,
                        SDVariable dl, SDVariable dr, SDVariable gradOut,
                        long rows, long cols) {
        super(sd, new SDVariable[]{aValues, aColIdx, aRowPtr, dl, dr, gradOut});
        addIArgument(rows, cols);
    }

    @Override
    public String opName() {
        return "csr_diag_mm_bp";
    }

    /**
     * Output data types:
     * <ol>
     *   <li>{@code dAValues} – same floating dtype as {@code aValues} (input index 0)</li>
     *   <li>{@code ddl}      – same floating dtype as {@code dl}      (input index 3)</li>
     *   <li>{@code ddr}      – same floating dtype as {@code dr}      (input index 4)</li>
     * </ol>
     */
    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes) {
        return Arrays.asList(dataTypes.get(0), dataTypes.get(3), dataTypes.get(4));
    }
}

File: libnd4j/include/ops/declarable/generic/sparse/csr_diag_mm_bp.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// csr_diag_mm_bp — backward pass for csr_diag_mm.
//
// Forward: out[e] = dl[i_e] * aValues[e] * dr[j_e]
//
// Backward (exact):
//   dAValues[e]  = gradOut[e] * dl[i_e] * dr[j_e]
//   ddl[i]       = sum_{e in row i}    gradOut[e] * aValues[e] * dr[j_e]
//   ddr[j]       = sum_{e: col j_e==j} gradOut[e] * aValues[e] * dl[i_e]
//
// Inputs (6 arrays):
//   [0] aValues   [nnz]     float  — original nonzero values of A
//   [1] aColIdx   [nnz]     int    — column indices of A
//   [2] aRowPtr   [rows+1]  int    — row pointers of A
//   [3] dl        [rows]    float  — left diagonal
//   [4] dr        [cols]    float  — right diagonal
//   [5] gradOut   [nnz]     float  — upstream gradient w.r.t. outValues
// IArgs:
//   [0] rows
//   [1] cols
// Outputs (3 arrays):
//   [0] dAValues  [nnz]     float  — gradient w.r.t. aValues
//   [1] ddl       [rows]    float  — gradient w.r.t. dl
//   [2] ddr       [cols]    float  — gradient w.r.t. dr
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_csr_diag_mm_bp)

#include <ops/declarable/headers/sparse.h>
#include <ops/declarable/helpers/sparse_diag_mm_bp.h>

namespace sd {
namespace ops {

CUSTOM_OP_IMPL(csr_diag_mm_bp, 6, 3, false, 0, 2) {
  auto aValues = INPUT_VARIABLE(0);
  auto aColIdx = INPUT_VARIABLE(1);
  auto aRowPtr = INPUT_VARIABLE(2);
  auto dl      = INPUT_VARIABLE(3);
  auto dr      = INPUT_VARIABLE(4);
  auto gradOut = INPUT_VARIABLE(5);

  const LongType rows = INT_ARG(0);
  const LongType cols = INT_ARG(1);

  // ── Validation ─────────────────────────────────────────────────────────────
  REQUIRE_TRUE(aValues->rankOf() == 1, 0,
               "csr_diag_mm_bp: aValues must be 1D, got rank %d", aValues->rankOf());
  REQUIRE_TRUE(aColIdx->rankOf() == 1, 0,
               "csr_diag_mm_bp: aColIdx must be 1D, got rank %d", aColIdx->rankOf());
  REQUIRE_TRUE(aRowPtr->rankOf() == 1, 0,
               "csr_diag_mm_bp: aRowPtr must be 1D, got rank %d", aRowPtr->rankOf());
  REQUIRE_TRUE(dl->rankOf() == 1, 0,
               "csr_diag_mm_bp: dl must be 1D, got rank %d", dl->rankOf());
  REQUIRE_TRUE(dr->rankOf() == 1, 0,
               "csr_diag_mm_bp: dr must be 1D, got rank %d", dr->rankOf());
  REQUIRE_TRUE(gradOut->rankOf() == 1, 0,
               "csr_diag_mm_bp: gradOut must be 1D, got rank %d", gradOut->rankOf());

  REQUIRE_TRUE(aValues->lengthOf() == aColIdx->lengthOf(), 0,
               "csr_diag_mm_bp: aValues and aColIdx must have the same length");
  REQUIRE_TRUE(aValues->lengthOf() == gradOut->lengthOf(), 0,
               "csr_diag_mm_bp: aValues and gradOut must have the same length");
  REQUIRE_TRUE(aRowPtr->lengthOf() == rows + 1, 0,
               "csr_diag_mm_bp: aRowPtr length must be rows+1=%lld, got %lld",
               (long long)(rows + 1), (long long)aRowPtr->lengthOf());
  REQUIRE_TRUE(dl->lengthOf() == rows, 0,
               "csr_diag_mm_bp: dl length must equal rows=%lld, got %lld",
               (long long)rows, (long long)dl->lengthOf());
  REQUIRE_TRUE(dr->lengthOf() == cols, 0,
               "csr_diag_mm_bp: dr length must equal cols=%lld, got %lld",
               (long long)cols, (long long)dr->lengthOf());

  REQUIRE_TRUE(aValues->dataType() == dl->dataType(), 0,
               "csr_diag_mm_bp: aValues and dl must have the same float dtype");
  REQUIRE_TRUE(aValues->dataType() == dr->dataType(), 0,
               "csr_diag_mm_bp: aValues and dr must have the same float dtype");
  REQUIRE_TRUE(aValues->dataType() == gradOut->dataType(), 0,
               "csr_diag_mm_bp: aValues and gradOut must have the same float dtype");
  REQUIRE_TRUE(aColIdx->dataType() == aRowPtr->dataType(), 0,
               "csr_diag_mm_bp: aColIdx and aRowPtr must have the same integer dtype");

  // ── Execute ────────────────────────────────────────────────────────────────
  auto dAValues = OUTPUT_NULLIFIED(0);
  auto ddl      = OUTPUT_NULLIFIED(1);
  auto ddr      = OUTPUT_NULLIFIED(2);

  sd::ops::helpers::csr_diag_mm_bp(*aValues, *aColIdx, *aRowPtr, *dl, *dr, *gradOut,
                                    *dAValues, *ddl, *ddr, rows, cols);

  return sd::Status::OK;
}

DECLARE_SHAPE_FN(csr_diag_mm_bp) {
  // dAValues shape == aValues shape  (same nnz, same float dtype)
  // ddl      shape == dl      shape  (rows)
  // ddr      shape == dr      shape  (cols)
  auto aValues = INPUT_VARIABLE(0);
  auto dl      = INPUT_VARIABLE(3);
  auto dr      = INPUT_VARIABLE(4);

  auto daShape  = ConstantShapeHelper::getInstance().createShapeInfo(
      aValues->dataType(), 'c', {aValues->lengthOf()});
  auto ddlShape = ConstantShapeHelper::getInstance().createShapeInfo(
      dl->dataType(), 'c', {dl->lengthOf()});
  auto ddrShape = ConstantShapeHelper::getInstance().createShapeInfo(
      dr->dataType(), 'c', {dr->lengthOf()});

  return SHAPELIST(daShape, ddlShape, ddrShape);
}

DECLARE_TYPES(csr_diag_mm_bp) {
  getOpDescriptor()
      ->setAllowedInputTypes(0, {ALL_FLOATS})   // aValues
      ->setAllowedInputTypes(1, {ALL_INTS})     // aColIdx
      ->setAllowedInputTypes(2, {ALL_INTS})     // aRowPtr
      ->setAllowedInputTypes(3, {ALL_FLOATS})   // dl
      ->setAllowedInputTypes(4, {ALL_FLOATS})   // dr
      ->setAllowedInputTypes(5, {ALL_FLOATS})   // gradOut
      ->setAllowedOutputTypes(0, {ALL_FLOATS})  // dAValues
      ->setAllowedOutputTypes(1, {ALL_FLOATS})  // ddl
      ->setAllowedOutputTypes(2, {ALL_FLOATS}); // ddr
}

}  // namespace ops
}  // namespace sd

#endif  // NOT_EXCLUDED(OP_csr_diag_mm_bp)
